core/src/main/java/org/apache/iceberg/BaseScan.java

```diff
@@ -143,6 +147,14 @@ protected ExecutorService planExecutor() {
     return context().planExecutor();
   }
 
+  protected Map<Integer, PartitionSpec> specsById() {
+    List<PartitionSpec> specs =
+        Lists.newArrayList(
+            Iterables.transform(
+                table.specs().values(), spec -> TableMetadata.updateSpecSchema(schema, spec)));
```

Comment:
What will happen if the spec is invalid for the given schema?
I see that the `TableMetadata` uses unchecked builds, but I was wondering if it is causing issues down the line.

Reply:
Specifications verification is not required for replacement by referring to the [4361](https://github.com/apache/iceberg/pull/4361)
  I think validation is needed here, so I added it.